File: src/events.ts
import { existsSync } from "node:fs";
import { resolve } from "node:path";
import { ValidationError } from "./errors.js";
import type { JsonObject } from "./models.js";
import { utcNow, validateIdentifier } from "./models.js";
import { Workspace } from "./storage.js";

export interface EventEnvelope { eventId: string; eventType: string; source: string; changeId: string; correlationId: string; payload: JsonObject; occurredAt: string }
export function createEvent(input: Omit<EventEnvelope, "occurredAt"> & { occurredAt?: string }): EventEnvelope {
  validateIdentifier(input.eventId, "event id", /^[A-Za-z0-9][A-Za-z0-9_-]*$/);
  for (const value of [input.eventType, input.source, input.changeId, input.correlationId]) if (!value.trim()) throw new ValidationError("event fields cannot be empty");
  return { ...input, occurredAt: input.occurredAt ?? utcNow() };
}
export class EventInbox {
  private readonly path: string;
  constructor(private readonly workspace: Workspace) { this.path = resolve(workspace.root, "events", "inbox"); }
  accept(event: EventEnvelope): boolean {
    this.workspace.initialize(); const record = resolve(this.path, `${event.eventId}.json`);
    if (existsSync(record)) { if (JSON.stringify(this.workspace.readJson(record)) !== JSON.stringify(event)) throw new ValidationError(`event id was reused with different content: ${event.eventId}`); return false; }
    this.workspace.writeJson(record, event as unknown as JsonObject); return true;
  }
}

File: src/cjhx_agile_workflow/events.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any

from .errors import ValidationError
from .models import utc_now, validate_identifier
from .storage import Workspace


@dataclass(frozen=True)
class EventEnvelope:
    event_id: str
    event_type: str
    source: str
    change_id: str
    correlation_id: str
    payload: dict[str, Any]
    occurred_at: str = field(default_factory=utc_now)

    def __post_init__(self) -> None:
        required = {
            "event_id": self.event_id,
            "event_type": self.event_type,
            "source": self.source,
            "change_id": self.change_id,
            "correlation_id": self.correlation_id,
        }
        missing = [name for name, value in required.items() if not value.strip()]
        if missing:
            raise ValidationError(f"event fields cannot be empty: {', '.join(missing)}")
        validate_identifier(
            self.event_id,
            "event id",
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_",
        )

    @classmethod
    def from_dict(cls, value: dict[str, Any]) -> EventEnvelope:
        mapping = {
            "eventId": "event_id",
            "eventType": "event_type",
            "source": "source",
            "changeId": "change_id",
            "correlationId": "correlation_id",
            "payload": "payload",
            "occurredAt": "occurred_at",
        }
        unknown = set(value) - set(mapping)
        if unknown:
            raise ValidationError(f"unknown event fields: {', '.join(sorted(unknown))}")
        normalized = {target: value[source] for source, target in mapping.items() if source in value}
        try:
            return cls(**normalized)
        except TypeError as error:
            raise ValidationError(f"invalid event envelope: {error}") from error

    def to_dict(self) -> dict[str, Any]:
        value = asdict(self)
        return {
            "eventId": value["event_id"],
            "eventType": value["event_type"],
            "source": value["source"],
            "changeId": value["change_id"],
            "correlationId": value["correlation_id"],
            "payload": value["payload"],
            "occurredAt": value["occurred_at"],
        }


class EventInbox:
    """Durable idempotency record for adapter/webhook consumers."""

    def __init__(self, workspace: Workspace) -> None:
        self.workspace = workspace
        self.path = workspace.root / "events" / "inbox"

    def accept(self, event: EventEnvelope) -> bool:
        self.workspace.initialize()
        record = self.path / f"{event.event_id}.json"
        if record.exists():
            existing = self.workspace.read_json(record)
            if existing != event.to_dict():
                raise ValidationError(
                    f"event id was reused with different content: {event.event_id}"
                )
            return False
        self.workspace.write_json(record, event.to_dict())
        return True
